vpr/test/test_noc_place_utils.cpp

```diff
@@ -0,0 +1,1666 @@
+            for (auto& link : golden_traffic_flow_routes[traffic_flow]) {
+                // update the link bandwidth in the NoC datastructure
+                double current_link_bandwidth = noc_ctx.noc_model.get_single_noc_link(link).get_bandwidth_usage();
+                noc_ctx.noc_model.get_single_mutable_noc_link(link).set_bandwidth_usage(current_link_bandwidth - curr_traffic_flow.traffic_flow_bandwidth);
+            }
+
+            // re-route the traffic flow
+            noc_ctx.noc_flows_router->route_flow(router_where_cluster_is_placed[curr_traffic_flow.source_router_cluster_id], router_where_cluster_is_placed[curr_traffic_flow.sink_router_cluster_id], golden_traffic_flow_routes[traffic_flow], noc_ctx.noc_model);
+
+            // go through the current traffic flow and reduce the bandwidths of the links (we only update this in the NoC, since these changes should be rectified by the test function)
+            // This shouldnt be updated in the golden bandwidths since we are imitating a swap of blocks and not having a real swap of blocks
+            for (auto& link : golden_traffic_flow_routes[traffic_flow]) {
+                // update the link bandwidth in the NoC datastructure
+                double current_link_bandwidth = noc_ctx.noc_model.get_single_noc_link(link).get_bandwidth_usage();
+                noc_ctx.noc_model.get_single_mutable_noc_link(link).set_bandwidth_usage(current_link_bandwidth + curr_traffic_flow.traffic_flow_bandwidth);
+                traffic_flow_route.push_back(link);
+            }
+
+            routed_traffic_flows.insert(traffic_flow);
+        }
+    }
+
+    // okay so now we undo the swapped blocks. We dont need to update the block locations in the placement datastructure since we initially never moved the blocks that were mswapped at the start.
+    // To undoe this we just need to update the noc link bandwidths as if there was no swap (we do this by calling the test function)
+    // This should then re-update the noc link bandwidths to their values before we imitated the swap above
+    // THe result is that the link bandwidths should match the golden link bandwidths that never changed after the initial router block placement (at a point before block swapping)
+    revert_noc_traffic_flow_routes(blocks_affected);
+
+    // now verify if the test function worked correctly by comparing the noc link bandwidths to the golden link bandwidths
+    int number_of_links = golden_link_bandwidths.size();
+    for (int link_number = 0; link_number < number_of_links; link_number++) {
+        NocLinkId current_link_id = (NocLinkId)link_number;
+        const NocLink& current_link = noc_ctx.noc_model.get_single_noc_link(current_link_id);
+
+        REQUIRE(golden_link_bandwidths[current_link_id] == current_link.get_bandwidth_usage());
+    }
+
+    delete routing_algorithm;
```

Comment:
Also very long. Any helper functions possible?

Reply:
Will leave as is for now. Not critical.